### PR TITLE
added VR phong Shading and new LUT

### PR DESCRIPTION
**Pull request by daron1337**
I've added the phong Shading to the VR.
You can enable/disable it setting the uniform uShading to 1 or 0.
You can easily define the intensity of ambient, specular, diffuse lights and shininess as float uniforms.
I've also added a muscle_bone LUT (the one used in medical image tools such as OSIRIX) and a full linear opacity LUT.
I think in the future we need to export LUTs in proper classes.

### CODE REVIEW

**Comment by NicolasRannou** [comment]
Do you have any idea why it gets dark when looking at a volume from the back?

**Comment by NicolasRannou** [comment]
maybe I'm confused and camera position doesn't change when we rotate the camera - will test a bite more

**Comment by daron1337** [comment]
@NicolasRannou Yes I think the problem is that camera position doesn't change. I'll check it

**Comment by NicolasRannou** [comment]
@daron1337 Sounds good - I'm looking at it too - also maybe the gradient calculation may not be correct or something is up with that:

The gradient is calculated in data space so maybe we have to transform it to world space to better results?

**Comment by daron1337** [comment]
Well, it could be! Nice catch.

**Comment by NicolasRannou** [comment]
For starters, gradient computation doesn't seem correct (needs works either at the trilinear interpolation probably...) I'm trying to display just the color of the gradients but not looking as expected and can not get it right yet somehow :/

**Comment by NicolasRannou** [comment]
OK, I finally 99% fixed the gradient computation. There is still some noise in the gradient somehow, will have to look into that at some point - will push the fix later today.

About the PR, suggestions:
1- Move all the shading code into 1 function.
2- We should probably expose all the light settings at the uniform/helpers levels (ambient, diffuse colors, shininess, etc.)
3- Maybe need a hook in case we want ambient/diffuse color to use "colorSample" as you did?

I want to clean up the WebGL code in general as it is getting messy, and one thing we considering is switching back to GLSLIFY.

Does it make sense?

**Comment by NicolasRannou** [comment]
No shading
screen shot 2017-11-30 at 10 10 38 am

**Comment by NicolasRannou** [comment]
Shading :)
screen shot 2017-11-30 at 10 10 52 am

**Comment by daron1337** [comment]
Hi @NicolasRannou ! wonderful.
Regarding the PR:
1- Move all the shading code into 1 function.
OK
2- We should probably expose all the light settings at the uniform/helpers levels (ambient, diffuse colors, shininess, etc.)
They already are in the shaders.vr.uniform.js, should we expose also ambient color and diffuse color in the uniform?

3- Maybe need a hook in case we want ambient/diffuse color to use "colorSample" as you did?

Should we set a uniform bool variable to be used for setting ambient/diffuse color to use "colorSample" if true? otherwise you can specify a specific color in the uniform

**Comment by NicolasRannou** [comment]
> should we expose also ambient color and diffuse color in the uniform?

Yes! :)

> Should we set a uniform bool variable to be used for setting ambient/diffuse color to use "colorSample" if true? otherwise you can specify a specific color in the uniform

Maybe if  "negative" `color.x` is provided we should just use the one from 'colorSample'?

**Comment by NicolasRannou** [comment]
Also the light position should probably also be exposed with a flag to automatically position light at camera for convenience -

**Comment by daron1337** [comment]
> Maybe if "negative" color.x is provided we should just use the one from 'colorSample'?

I don't know, I think it could be misleading maybe it's better to have a dedicated variable in the uniform. Where do you set color.x?

**Comment by NicolasRannou** [comment]
yes you are right, let's add another variable

**Comment by NicolasRannou** [comment]
FYI about making things work in a function, I have been playing with this one which also looked good before: https://www.shadertoy.com/view/lt33z7

(from http://jamie-wong.com/2016/07/15/ray-marching-signed-distance-functions/)

**Comment by NicolasRannou** [comment]
Couple more things - 
I pushed the correct gradient computation so if you update AMI you show notice the difference.

Couple of things have to be updated in your PR:

1- Move gradient from local to world space:
```
// to add in vr.fragment (ideally somewhere else in its own shader)
// function to invert matrix
mat4 inverse(mat4 m) {
  float
      a00 = m[0][0], a01 = m[0][1], a02 = m[0][2], a03 = m[0][3],
      a10 = m[1][0], a11 = m[1][1], a12 = m[1][2], a13 = m[1][3],
      a20 = m[2][0], a21 = m[2][1], a22 = m[2][2], a23 = m[2][3],
      a30 = m[3][0], a31 = m[3][1], a32 = m[3][2], a33 = m[3][3],

      b00 = a00 * a11 - a01 * a10,
      b01 = a00 * a12 - a02 * a10,
      b02 = a00 * a13 - a03 * a10,
      b03 = a01 * a12 - a02 * a11,
      b04 = a01 * a13 - a03 * a11,
      b05 = a02 * a13 - a03 * a12,
      b06 = a20 * a31 - a21 * a30,
      b07 = a20 * a32 - a22 * a30,
      b08 = a20 * a33 - a23 * a30,
      b09 = a21 * a32 - a22 * a31,
      b10 = a21 * a33 - a23 * a31,
      b11 = a22 * a33 - a23 * a32,

      det = b00 * b11 - b01 * b10 + b02 * b09 + b03 * b08 - b04 * b07 + b05 * b06;

  return mat4(
      a11 * b11 - a12 * b10 + a13 * b09,
      a02 * b10 - a01 * b11 - a03 * b09,
      a31 * b05 - a32 * b04 + a33 * b03,
      a22 * b04 - a21 * b05 - a23 * b03,
      a12 * b08 - a10 * b11 - a13 * b07,
      a00 * b11 - a02 * b08 + a03 * b07,
      a32 * b02 - a30 * b05 - a33 * b01,
      a20 * b05 - a22 * b02 + a23 * b01,
      a10 * b10 - a11 * b08 + a13 * b06,
      a01 * b08 - a00 * b10 - a03 * b06,
      a30 * b04 - a31 * b02 + a33 * b00,
      a21 * b02 - a20 * b04 - a23 * b00,
      a11 * b07 - a10 * b09 - a12 * b06,
      a00 * b09 - a01 * b07 + a02 * b06,
      a31 * b01 - a30 * b03 - a32 * b00,
      a20 * b03 - a21 * b01 + a22 * b00) / det;
}

// before the ray marching create the invert 'dataToWorld' matrix:
...
  float accumulatedAlpha = 0.0;
  mat4 dataToWorld = inverse(uWorldToData);

  for(int rayStep = 0; rayStep < maxSteps; rayStep++){
...

// map gradient to world space and normalize before using
// no need of "-" sign anymore
// we avoid to call "normalize" as it may be undefined if vector length == 0.
gradient= (vec3(dataToWorld * vec4(gradient, 0.)));
if (length(gradient) > 0.0) {
  gradient = normalize(gradient);
}

```

**Comment by NicolasRannou** [comment]
That should be it

File: src/shaders/shaders.vr.fragment.js [hunk 1]
```diff
@@ -56,11 +56,46 @@ void getIntensity(in vec3 dataCoordinates, out float intensity, out vec3 gradien
   intensity = ( intensity - windowMin ) / uWindowCenterWidth[1];
 }
 
+void phongShading(in vec3 lightOrigin, in vec3 rayOrigin, in vec3 currentPosition, in vec3 gradient, in vec4 colorSample, out vec3 litColor){
+  vec3 normal = -normalize(gradient);
```
Comment (NicolasRannou) [inline src/shaders/shaders.vr.fragment.js]: nothing to here anymore  since gradient will already well oriented after suggestion from the PR thread

**Comment by daron1337** [comment]
Perfect! Now the light seems correct and the update to gradient computation is awesome!

screen shot 2017-12-01 at 10 13 02

**Comment by NicolasRannou** [comment]
Thanks! In the coming weeks my plan it to investigate/go back to GLSLIFY to have much cleaner/reusable shader code - got to think about it!

**Comment by NicolasRannou** [comment]
@daron1337 hmm somehow the results are not quite as clean as I'd hope, is it easy for you to test a branch on your dataset if I create one?

**Comment by NicolasRannou** [comment]
That does look nicer on my test data, can you confirm on yours?

https://github.com/FNNDSC/ami/tree/phong

**Comment by NicolasRannou** [comment]
also I inverted the behavior of "uSampleColorToAmbient", "uSampleColorToDiffuse" and "uLightPositionInCamera":
0 means false -> do not use color sample for ambient/diffuse
1 means true -> use

**Comment by NicolasRannou** [comment]
@daron1337 I did update the code, please let me know if that also looks better for you! Thanks again for the contribution that does look great, really happy with results!

**Comment by daron1337** [comment]
It looks definitely better! However after your last update I have again the problem of "zooming into the cube" now.  Do you have any suggestion?

**Comment by NicolasRannou** [comment]
Hmm does it get "transparent" or black? Can you try with shading off?

**Comment by NicolasRannou** [comment]
also if that is OK, can new share a screenshot of the new shading so we have some kind of history of the changes in this thread?

**Comment by NicolasRannou** [comment]
hmm actually with shading on it is all black
screen shot 2017-12-01 at 4 51 01 pm

**Comment by daron1337** [comment]
With shading ON while zooming into the cube is all black (like before the fix of a week ago). Before your last update it was still working :) 
However now is definitely better.
screen shot 2017-12-01 at 16 51 28

**Comment by NicolasRannou** [comment]
so one way to fix it seems to only apply phong shading if component is opaque enough:
```
    if (uShading == 1 && uInterpolation != 0 && alphaSample > .3) {
```

But I am not sure which is the right (if any right) approach is.

Another approach to test which seems more correct is to actually "normalize" the gradient by the window width (or 2*window width). This way is there is a small gradient the ambient, diffuse and specular contributions will be negligible.

**Comment by NicolasRannou** [comment]
Been thinking/testing a a bit more about it and the problem seems to be that something is off with the last part of the phong computation:
```
return k_a + lightIntensity * (k_d * dotLN  + k_s * pow(dotRV, shininess));
```

In some cases it doesn't work (not 100% sure why yet.)

If you replace it by: `return vec3(0., 0., 1.);`, it lets you zoom into the volume properly.

Can you confirm?

Thanks @daron1337 !

**Comment by NicolasRannou** [comment]
Seems Bling-Phong works better: https://en.wikipedia.org/wiki/Blinn%E2%80%93Phong_shading_model

**Comment by daron1337** [comment]
> If you replace it by: return vec3(0., 0., 1.);, it lets you zoom into the volume properly.

Yes, but in this way the computed color it seems wrong... Any idea?

**Comment by NicolasRannou** [comment]
Will push something soon, got something working :) (in a meeting)

**Comment by NicolasRannou** [comment]
@daron1337 not perfect but there are couple of things to be tried out :/

**Comment by NicolasRannou** [comment]
There is the real fix @daron1337 : https://github.com/FNNDSC/ami/commit/a3bbdf75a93fcaf2b51dd2323f73b85b95366a81

Cheers

**Comment by daron1337** [comment]
Thanks!

**Pull request merged**